### PR TITLE
DAS sampling on sync

### PR DESCRIPTION
**Pull request by dapplion**
## Issue Addressed

Implements the foundational networking infrastructure / wiring for sampling.

Consumers will trigger sampling sending a `SyncMessage::SampleBlock(Hash256)`.  Sampling requests are tracked independently on `SyncManager::sampling`. Each `ActiveSamplingRequest` manages multiple individual sample column requests. A `ActiveSamplingRequest` manages retries and may extend the count of sample columns following LossyDAS. When a sampling requests has enough required successes, it's complete.

SyncManager will forward the sampling result (a single bit Ok/Not) to the chain who should update the corresponding fork-choice node with trailing DA.

**Range sync**

Range sync may explicitly wait for the block to be sampled before sending it for import, or sampling could trail behind imported blocks. This PR can easily accommodate both, by adding a new enum variant `SamplingRequester::RangeSync`.

### CODE REVIEW

File: beacon_node/beacon_processor/src/lib.rs
Comment (jimmygchen) [inline beacon_node/beacon_processor/src/lib.rs:847]: Looks like we never read / `pop` from these queues. We'll need to add them under this section:
https://github.com/sigp/lighthouse/blob/9de3178c7ffe6939210c643e28ec35d154d0c679/beacon_node/beacon_processor/src/lib.rs#L992

Comment (jimmygchen) [reply beacon_node/beacon_processor/src/lib.rs:847]: Also worth adding the queue metrics so we don't forget:
https://github.com/sigp/lighthouse/blob/9de3178c7ffe6939210c643e28ec35d154d0c679/beacon_node/beacon_processor/src/lib.rs#L1420-L1423

File: beacon_node/lighthouse_network/src/rpc/methods.rs
Comment (jimmygchen) [inline beacon_node/lighthouse_network/src/rpc/methods.rs]:
```suggestion
    pub fn new(data_column_ids: Vec<DataColumnIdentifier>, spec: &ChainSpec) -> Self {
```

File: beacon_node/network/src/router.rs
Comment (jimmygchen) [inline beacon_node/network/src/router.rs:519]: Maybe worth keeping the `peer_id` logging here?

File: beacon_node/network/src/router.rs
Comment (jimmygchen) [inline beacon_node/network/src/router.rs:581]: Same as above, should we keep `peer_id`?

File: beacon_node/network/src/sync/manager.rs
Comment (jimmygchen) [inline beacon_node/network/src/sync/manager.rs:129]: Please add some doc here

File: beacon_node/network/src/sync/manager.rs
Comment (jimmygchen) [inline beacon_node/network/src/sync/manager.rs:366]: Perhaps rename this to `SingleDataColumn` for consistency?

File: beacon_node/network/src/sync/manager.rs
Comment (jimmygchen) [inline beacon_node/network/src/sync/manager.rs:985]: We should handle `RangeBlockAndDataColumns` here instead right?

File: beacon_node/network/src/sync/manager.rs
Comment (jimmygchen) [inline beacon_node/network/src/sync/manager.rs]: Are you still thinking to add a log here? It looks like we don't log anything if block or blob requests aren't found.

File: beacon_node/network/src/sync/network_context.rs
Comment (jimmygchen) [inline beacon_node/network/src/sync/network_context.rs]: Should be able to use `enr.custody_subnet_count()` for this now.

https://github.com/sigp/lighthouse/blob/d5e7e73abc64312d7a6f5ce3a54b0df455e2b243/beacon_node/lighthouse_network/src/discovery/enr.rs#L67

File: beacon_node/network/src/sync/network_context.rs
Comment (jimmygchen) [inline beacon_node/network/src/sync/network_context.rs:583]: It feels a bit confusing that the tuple order (ID, Resp) is the opposite to the generic parameters (Resp, ID)

File: beacon_node/network/src/sync/sampling.rs
Comment (jimmygchen) [inline beacon_node/network/src/sync/sampling.rs:25]: Does this mean we're only triggering sampling after importing a block? What other variants are you thinking to add?

File: beacon_node/network/src/sync/sampling.rs
Comment (jimmygchen) [inline beacon_node/network/src/sync/sampling.rs]: Should we be using `EthSpec::number_of_columns()` instead? (currently `128`)

**Comment by jimmygchen** [comment]
Could you merge latest `das` onto your branch? I fixed a merge conflict issue and it should now trigger the test suite again.

File: beacon_node/network/src/sync/sampling.rs
Comment (jimmygchen) [inline beacon_node/network/src/sync/sampling.rs]: lint would probably complain about this, we should use `get` here to prevent index out of bounds.

File: beacon_node/network/src/sync/sampling.rs
Comment (jimmygchen) [inline beacon_node/network/src/sync/sampling.rs:211]: is it worth adding an extra check that `required_successes` is always less than the length of `column_shuffle`?

File: beacon_node/network/src/sync/network_context.rs
Comment (jimmygchen) [inline beacon_node/network/src/sync/network_context.rs]: `peers` returns all known peers in the `PeerDB`. Maybe `connected_peers` should be used here instead?

File: beacon_node/network/src/sync/sampling.rs
Comment (jimmygchen) [inline beacon_node/network/src/sync/sampling.rs:240]: Can we add some docs for the return value? as `Option<()` isn't very descriptive.

File: beacon_node/network/src/sync/sampling.rs
Comment (jimmygchen) [inline beacon_node/network/src/sync/sampling.rs:296]: Do we currently or plan to keep track of the peers that fail to return sample results somewhere and penalise if it's recurring?

File: beacon_node/network/src/sync/sampling.rs
Comment (jimmygchen) [inline beacon_node/network/src/sync/sampling.rs]: Maybe some reasons on why its removed would be useful, e.g. "Remove completed sampling request" (completed means success/failed here)

**Comment by jimmygchen** [comment]
Hi @dapplion 

The overall structure of looks really good and simple, I like it! 

From my understanding, this PR does not implement triggering of sampling block`SyncMessage::SampleBlock` yet. What's your thoughts about this?  We could potentially trigger this from `BeaconProcessor` as soon as we receive the first verified block component?

I've added some comments above, and additionally I think it would be useful to:
- Add some docs to the pub methods of `ActiveSamplingRequest`
- Add some metrics to track sampling, maybe:
  - sample request attempts
  - sample request success
  - sample request failures
  - active sampling requests

I don't think we need to address all of my comments in the PR, happy to merge this PR as soon as it's in a good shape for merge, and we've tracked the extra tasks in the tracking issue.

Nice work! 🔥

**Review by jimmygchen — CHANGES_REQUESTED** [review]
Added my review above.

Some diagrams I've created to help my understanding, may helps other reviewers:


### Initiating sampling request

```mermaid
sequenceDiagram
    Note over NetworkBeaconProcessor: On slot OR component received
    NetworkBeaconProcessor->>SyncManager: `SampleBlock(Hash256, Slot)` message
    SyncManager->>Sampling: new sample request
    Note over Sampling: create `ActiveSamplingRequest` for block root 
    Note over Sampling: sends request via `SyncNetworkContext`
```

### Handling sampling RPC response

```mermaid
sequenceDiagram
    Note over SyncManager: On RPC column received
    SyncManager->>Sampling: downloaded column response
    Sampling->>BeaconProcessor: send for verification (`Work::RpcVerifyDataColumn`)
    Note over BeaconProcessor: validate data columns
    BeaconProcessor->>SyncManager: send `SampleVerified` message
    SyncManager->>Sampling: send result for sampling status re-evaluation
    Sampling-->>SyncManager: returns sampling result
    SyncManager->>BeaconProcessor: Send sampling completed (`Work::SamplingResult`)
    BeaconProcessor->>BeaconChain: Update fork choice (Unimplemented)
```

File: beacon_node/network/src/router.rs
Comment (dapplion) [reply beacon_node/network/src/router.rs:519]: I don't think so, an error is an implementation / internal error on our end. There's nothing a peer can do to cause this error

File: beacon_node/network/src/sync/manager.rs
Comment (dapplion) [reply beacon_node/network/src/sync/manager.rs:366]: So `SingleBlob` blob does not really mean 1 blob, it means a blob for a SingleLookup. Here data columns are not strictly part of a SingleLookup as range sync will also consume then. I can do a PR latter to align `SingleBlob` into `BlobByRoot` which is also correct, but using `SingleDataColumn` would be a missnomer.

File: beacon_node/network/src/sync/manager.rs
Comment (dapplion) [reply beacon_node/network/src/sync/manager.rs:985]: I think we should rename `RangeBlockAndBlobs` to `RangeBlockComponents` or `RangeBlockAndData` to be represent the set of sidecars the block needs. There's no need to have a separate enum variant. See
- https://github.com/sigp/lighthouse/pull/5561

File: beacon_node/network/src/sync/manager.rs
Comment (dapplion) [reply beacon_node/network/src/sync/manager.rs]: Fixed, None is an expected value and should not have a dedicated log

File: beacon_node/network/src/sync/sampling.rs
Comment (dapplion) [reply beacon_node/network/src/sync/sampling.rs:25]: We could trigger sampling as soon as we see the block over gossip for a not yet imported block. Also if we are doing same slot sampling.

File: beacon_node/network/src/sync/sampling.rs
Comment (dapplion) [reply beacon_node/network/src/sync/sampling.rs:211]: I don't think so, the list of successes is hardcoded by us. It is only dynamic to control testing which we also control. This way we don't need to make the constructor fallible.

File: beacon_node/network/src/sync/sampling.rs
Comment (dapplion) [reply beacon_node/network/src/sync/sampling.rs:296]: Scoring in sampling is a complicated topic. There's no clear attribution if the proposer is playing delay games to force donwscoring events or the peers are being actively malicious. For now I would just do nothing and figure it out latter in the interop

**Comment by dapplion** [comment]
> From my understanding, this PR does not implement triggering of sampling blockSyncMessage::SampleBlock yet. What's your thoughts about this? We could potentially trigger this from BeaconProcessor as soon as we receive the first verified block component?

The easiest now is to wait for the block to make sure if has > 0 blobs.

> Add some docs to the pub methods of ActiveSamplingRequest

Done

> Add some metrics to track sampling

WIP

**Comment by jimmygchen** [comment]
> The easiest now is to wait for the block to make sure if has > 0 blobs.

If we receive a `DataColumnSidecar` which would also have list of kzg commitments, would that not be sufficient? or is there any security concerns around triggering sampling before we get a block?

File: beacon_node/beacon_processor/src/lib.rs
Comment (jimmygchen) [inline beacon_node/beacon_processor/src/lib.rs:1007]: I think this is good since we prioritise importing rpc blocks over everything else 👍

File: beacon_node/beacon_processor/src/lib.rs
Comment (jimmygchen) [inline beacon_node/beacon_processor/src/lib.rs]:
```suggestion
                    rpc_verify_data_column_queue.len() as i64,
```

File: beacon_node/beacon_processor/src/lib.rs
Comment (jimmygchen) [inline beacon_node/beacon_processor/src/lib.rs]:
```suggestion
                    sampling_result_queue.len() as i64,
```

File: beacon_node/beacon_processor/src/metrics.rs
Comment (jimmygchen) [inline beacon_node/beacon_processor/src/metrics.rs]:
```suggestion
        "beacon_processor_sampling_result_queue_total",
```

**Comment by dapplion** [comment]
> If we receive a DataColumnSidecar which would also have list of kzg commitments, would that not be sufficient? or is there any security concerns around triggering sampling before we get a block?

Touche, not more complains from me

File: beacon_node/network/src/sync/network_context.rs
Comment (jimmygchen) [inline beacon_node/network/src/sync/network_context.rs:173]: The method already fallsback to min custody requirement. An error here means there's an issue decoding ENR, I'm not sure if we should fallback to min custody requirement here? I think we should at least log an error here to avoid a hard to find bug.

I think we could probably fallback to `min_custody_requirment` though if ENR fails to decode, providing that we do the same everywhere else.

**Comment by dapplion** [comment]
@jimmygchen Added metrics on https://github.com/sigp/lighthouse/pull/5616/commits/f8acdac4c4dc4cb54ebaa5d3871302f0c78bd6c6 bundled with remote marge by mistake

File: beacon_node/network/src/sync/sampling.rs
Comment (jimmygchen) [inline beacon_node/network/src/sync/sampling.rs:490]: Maybe we could use status messages, similar to how we do it in range_sync - happy to leave this out of this current PR though

https://github.com/sigp/lighthouse/blob/23671531bcb2c8050888b51b80f1be2310384037/beacon_node/network/src/sync/manager.rs#L345-L351

Comment (dapplion) [reply beacon_node/network/src/sync/sampling.rs:490]: It's still tricky because peers can have slightly different heads on the same fork such that they _should_ be grouped together but they will appear in different forks

Comment (jimmygchen) [reply beacon_node/network/src/sync/sampling.rs:490]: Yep let's think about this - going to resolve this and we'll get back to this TODO later

File: beacon_node/network/src/sync/network_context.rs
Comment (jimmygchen) [reply beacon_node/network/src/sync/network_context.rs:173]: Had a chat with Lion offline and it makes sense to ignore peer ENR decoding error:

> RE ENR decode, can decoding fail due to the other party creating a bad ENR? In that case we should ignore the value

File: beacon_node/network/src/sync/sampling.rs
Comment (jimmygchen) [reply beacon_node/network/src/sync/sampling.rs:296]: Sounds good

**Comment by jimmygchen** [comment]
Nice, we're close to merging this! Just the last few metrics comments and a green `network-tests` and we're good to merge!

Ignore the `execution-engine-integtration` as it's been fixed on `unstable` - will merge `unstable` into `das` after this PR is merged.

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (jimmygchen) [inline beacon_node/beacon_chain/src/beacon_chain.rs:2862]: I think it would be possible that sampling is completed before block is imported into fork choice - we'll need to remember handling this scenario.

File: beacon_node/network/src/sync/sampling.rs
Comment (jimmygchen) [inline beacon_node/network/src/sync/sampling.rs:77]: I think the main one we want to avoid retrying is `TooManyFailures` right? so it doesn't reset the required success to the min number? ([`LossyDAS`](https://ethresear.ch/t/lossydas-lossy-incremental-and-diagonal-sampling-for-data-availability/18963#incrementaldas-dynamically-increase-the-sample-size-10))

File: beacon_node/network/src/network_beacon_processor/gossip_methods.rs
Comment (jimmygchen) [inline beacon_node/network/src/network_beacon_processor/gossip_methods.rs]: Minor - for supernodes, we might get flooded with `Received SampleBlock message` if we're sending a sync message for every gossip column received
https://github.com/sigp/lighthouse/blob/f8acdac4c4dc4cb54ebaa5d3871302f0c78bd6c6/beacon_node/network/src/sync/manager.rs#L656

File: beacon_node/network/src/sync/network_context/requests.rs
Comment (jimmygchen) [inline beacon_node/network/src/sync/network_context/requests.rs]: `data_column`

File: beacon_node/network/src/network_beacon_processor/gossip_methods.rs
Comment (dapplion) [reply beacon_node/network/src/network_beacon_processor/gossip_methods.rs]: Right, triggering on column is a timing optimization. We can drop it for simplicity

File: beacon_node/network/src/sync/sampling.rs
Comment (jimmygchen) [reply beacon_node/network/src/sync/sampling.rs:77]: Resolving this as we can address later.

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (jimmygchen) [reply beacon_node/beacon_chain/src/beacon_chain.rs:2862]: I'll add a comment

**Comment by jimmygchen** [comment]
I've pushed a commit to resolve some minor review comments, and added a check for `PEER_DAS_EPOCH` before sending a sampling request.

I've looked into the test failures (network_tests, fallback sim, and basic sim tests), and ~~they all passed locally, so I'll let it run on CI again (could be caching issue).~~

**EDIT**: found a `Duration` subtraction panic in the logs, investigating now.

I'll do a merge from `unstable` to `das` after this is merged to `das`.

**Comment by jimmygchen** [comment]
I couldn't reproduce the sim tests issues locally, and it appears to be happening on [other PRs](https://github.com/sigp/lighthouse/actions/runs/8859873093) too. I'm going to look into it later tonight.

The error I'm seeing on CI is:
> 2024-04-29T07:35:16.5280075Z thread 'main' panicked at /rustc/25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04/library/core/src/time.rs:954:31:
2024-04-29T07:35:16.5280267Z overflow when subtracting durations
2024-04-29T07:35:16.5280709Z note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

The execution tests should pass once we merge `unstable` into `das`, so I think we're good to merge this one now!

**Pull request merged**
